jobsdb/admin.go

```diff
@@ -135,3 +139,111 @@ func (jd *HandleT) failExecutingDSInTx(txHandler transactionHandler, ds dataSetT
 	)
 	return err
 }
+
+func (jd *HandleT) startCleanupLoop(ctx context.Context) {
+	jd.backgroundGroup.Go(misc.WithBugsnag(func() error {
+		jd.oldJobsCleanupRoutine(ctx)
+		return nil
+	}))
+}
+
+func (jd *HandleT) oldJobsCleanupRoutine(ctx context.Context) {
+	jd.doCleanupOldJobs(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-jd.TriggerJobCleanUp():
+			jd.doCleanupOldJobs(ctx)
+		}
+	}
+}
+
+func (jd *HandleT) doCleanupOldJobs(ctx context.Context) {
+	var (
+		queryUnprocessed = true
+		queryProcessed   = true
+
+		unprocessedAfterJobID *int64
+		processedAfterJobID   *int64
+
+		batchSize            = config.GetInt("jobsdb.cleanupBatchSize", 100)
+		cleanupRetryInterval = config.GetDuration(
+			"jobsdb.cleanupRetryInterval",
+			10, time.Second,
+		)
+	)
+
+	for {
+		jobsToCleanup := make([]*JobT, 0)
+		if queryUnprocessed {
+			unprocessed, err := jd.GetUnprocessed(ctx, GetQueryParamsT{
+				IgnoreCustomValFiltersInQuery: true,
+				JobsLimit:                     batchSize,
+				AfterJobID:                    unprocessedAfterJobID,
+			})
+			if err != nil {
+				jd.logger.Errorf("error getting unprocessed jobs to cleanup: %w", err)
+				time.Sleep(cleanupRetryInterval)
+				continue
+			}
+			if len(unprocessed.Jobs) > 0 {
+				unprocessedAfterJobID = &(unprocessed.Jobs[len(unprocessed.Jobs)-1].JobID)
+				unprocessedJobsToCleanup := lo.Filter(unprocessed.Jobs, func(job *JobT, _ int) bool {
+					return job.CreatedAt.Before(time.Now().Add(-jd.JobMaxAge))
+				})
+				jobsToCleanup = append(jobsToCleanup, unprocessedJobsToCleanup...)
+				if len(unprocessedJobsToCleanup) < batchSize {
+					queryUnprocessed = false
+				}
+			} else {
+				queryUnprocessed = false
+			}
+		}
+
+		if queryProcessed {
+			processed, err := jd.GetProcessed(ctx, GetQueryParamsT{
+				IgnoreCustomValFiltersInQuery: true,
+				JobsLimit:                     batchSize,
+				AfterJobID:                    processedAfterJobID,
+			})
+			if err != nil {
+				jd.logger.Errorf("error getting processed jobs to cleanup: %w", err)
+				time.Sleep(cleanupRetryInterval)
```

Comment:
sleep needs to respect the context

Reply:
please elaborate
I didn't understand this

Reply:
we have a function that respects the context, i.e. cancels the sleep if the context is cancelled. I think it is called sleepCtx or something similar